src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_expert_opt.cpp

```diff
@@ -0,0 +1,1215 @@
+            bin_post_id = 1;
+            post_op_silu();
+            post_op_bin_mul(true);
+        }
+
+        create(eng);
+    }
+};
+
+// all jit-based/performance-aware function should be a functor/callable because:
+//   - it needs to hold reference to kernel (to save build time & resources)
+//   - it needs to do other compile time preparation work and hold the relevant
+//     runtime-data-struct (to make runtime faster)
+// to optimize compile-time-workload itself, the functor instance itself should be
+// cached with compile-time parameter as the key.
+//
+// because it's a functor, which supposed to have no states, so cache-factory should
+// always return shared_ptr to constant object, so it won't behave differently when being
+// called by different caller, and this also ensure it's multi-threading safe since it
+// won't modify it's content.
+//
+template <typename... TTypes>
+class tuple_hasher {
+private:
+    typedef std::tuple<TTypes...> Tuple;
+    template <int N>
+    size_t hash(Tuple& value) const {
+        return 0;
+    }
+    template <int N, typename THead, typename... TTail>
+    size_t hash(Tuple& value) const {
+        constexpr int Index = N - sizeof...(TTail) - 1;
+        return std::hash<THead>()(std::get<Index>(value)) ^ hash<N, TTail...>(value);
+    }
+
+public:
+    size_t operator()(Tuple value) const {
+        auto hv = hash<sizeof...(TTypes), TTypes...>(value);
+        return hv;
+    }
+};
+
+// create const object with internal cache with constructor-args as the key
+// this helps reduces construction time overhead, and perfectly suitable
+// for caching functor/callable.
+template <class T, typename... CArgs>
+std::shared_ptr<const T> make_cacheable(dnnl::engine eng, CArgs... cargs) {
+    std::shared_ptr<const T> sptr;
+    auto key = std::make_tuple(cargs...);
+    static std::unordered_map<decltype(key), std::weak_ptr<const T>, tuple_hasher<CArgs...>> cache;
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> guard(mutex);
+    auto it = cache.find(key);
+    if (it != cache.end()) {
+        auto& wptr = it->second;
+        sptr = wptr.lock();
+        if (!sptr) {
+            sptr = std::make_shared<T>(eng, cargs...);
+            // ECOUT("make_cacheable re-constructed: ", typeid(T).name(), "(", cargs..., ")");
+            wptr = sptr;
+        }
+    } else {
+        sptr = std::make_shared<T>(eng, cargs...);
+        // ECOUT("make_cacheable constructed: ", typeid(T).name(), "(", cargs..., ")");
+        cache.emplace(std::make_pair(key, std::weak_ptr<const T>(sptr)));
+    }
+    return sptr;
+}
+
+struct onednn_linear {
+    std::shared_ptr<const onednn_matmul> mm;
+    dnnl::memory weight;
+    dnnl::memory scale;
+    dnnl::memory zp;
+    dnnl::matmul m_prim;
+    dnnl::memory::dim m_K;
+    dnnl::memory::dim m_N;
+    dnnl::memory::dim m_batch;
+    dnnl::memory::data_type m_a_type;
+    int bin_post_id;
+
+    static onednn_linear create(dnnl::engine eng,
```

Comment:
I think we should reuse existing oneDNN FC/MatMul implementations instead of using oneDNN interfaces here explicitly